File: src/static/js/validation.js
/**
 * @fileoverview Frontend validation utilities for the Blackjack Card Counter.
 * @module validation
 * @description Provides comprehensive validation functions for user inputs,
 * game state, and API responses to ensure data integrity and prevent errors.
 */

/**
 * Represents a playing card rank (e.g., 'A', '2', ..., '10', 'J', 'Q', 'K')
 * @typedef {('A'|'2'|'3'|'4'|'5'|'6'|'7'|'8'|'9'|'10'|'J'|'Q'|'K')} CardRank
 */

/**
 * Represents a counting system identifier
 * @typedef {('hiLo'|'ko'|'omega2'|'zenCount')} CountingSystem
 */

/**
 * Standard validation result object
 * @typedef {Object} ValidationResult
 * @property {boolean} isValid - Whether the validation passed
 * @property {string} [message] - Error message if validation failed
 */

/**
 * Valid card ranks for the game
 * @type {Set<CardRank>}
 * @readonly
 */
const VALID_CARDS = new Set(['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']);

/**
 * Supported card counting systems
 * @type {Set<CountingSystem>}
 * @readonly
 */
const VALID_COUNTING_SYSTEMS = new Set(['hiLo', 'ko', 'omega2', 'zenCount']);

/**
 * Validates if a string represents a valid playing card rank.
 * @param {unknown} card - The value to validate as a card rank
 * @returns {card is CardRank} True if the input is a valid card rank
 * @example
 * // Returns true
 * isValidCard('A');
 * @example
 * // Returns false
 * isValidCard('X');
 */
function isValidCard(card) {
    return typeof card === 'string' && VALID_CARDS.has(/** @type {CardRank} */ (card.toUpperCase()));
}

/**
 * Validates an array of card ranks.
 * @param {unknown} cards - The value to validate as an array of card ranks
 * @returns {ValidationResult} Validation result object
 * @example
 * // Returns { isValid: true }
 * validateCards(['A', 'K', '10']);
 * @example
 * // Returns { isValid: false, message: 'Cards must be provided as an array' }
 * validateCards('not an array');
 * @example
 * // Returns { isValid: false, message: 'Invalid card at position 2: X. Valid values are: ...' }
 * validateCards(['A', 'K', 'X']);
 */
function validateCards(cards) {
    if (!Array.isArray(cards)) {
        return {
            isValid: false,
            message: 'Cards must be provided as an array'
        };
    }

    // Handle empty array case
    if (cards.length === 0) {
        return {
            isValid: true
        };
    }

    // Check each card in the array
    for (const [index, card] of cards.entries()) {
        if (!isValidCard(card)) {
            return {
                isValid: false,
                message: `Invalid card at position ${index + 1}: "${card}". ` +
                         `Valid values are: ${Array.from(VALID_CARDS).join(', ')}`
            };
        }
    }

    return { isValid: true };
}

/**
 * Validates the number of decks in a blackjack shoe.
 * @param {unknown} decks - The value to validate as number of decks
 * @returns {ValidationResult} Validation result object
 * @example
 * // Returns { isValid: true }
 * validateDecks(6);
 * @example
 * // Returns { isValid: false, message: 'Number of decks must be an integer between 1 and 10' }
 * validateDecks(0);
 */
function validateDecks(decks) {
    const numDecks = Number(decks);
    if (isNaN(numDecks) || !Number.isInteger(numDecks) || numDecks < 1 || numDecks > 10) {
        return {
            isValid: false,
            message: 'Number of decks must be an integer between 1 and 10'
        };
    }
    return { isValid: true };
}

/**
 * Validates a card counting system identifier.
 * @param {unknown} system - The counting system to validate
 * @returns {ValidationResult & { system?: CountingSystem }} Validation result with optional valid system
 * @example
 * // Returns { isValid: true, system: 'hiLo' }
 * validateCountingSystem('hiLo');
 * @example
 * // Returns { isValid: false, message: 'Invalid counting system: invalid. Valid systems are: ...' }
 * validateCountingSystem('invalid');
 */
function validateCountingSystem(system) {
    if (typeof system !== 'string' || !VALID_COUNTING_SYSTEMS.has(/** @type {CountingSystem} */ (system))) {
        return {
            isValid: false,
            message: `Invalid counting system: ${system}. ` +
                    `Valid systems are: ${Array.from(VALID_COUNTING_SYSTEMS).join(', ')}`
        };
    }
    return { 
        isValid: true,
        system: /** @type {CountingSystem} */ (system)
    };
}

/**
 * Validates the deck penetration percentage.
 * @param {unknown} penetration - The penetration percentage to validate (0-100)
 * @returns {ValidationResult} Validation result object
 * @example
 * // Returns { isValid: true }
 * validatePenetration(75);
 * @example
 * // Returns { isValid: false, message: 'Penetration must be a number between 0 and 100' }
 * validatePenetration(150);
 */
function validatePenetration(penetration) {
    const penValue = Number(penetration);
    if (isNaN(penValue) || penValue < 0 || penValue > 100) {
        return {
            isValid: false,
            message: 'Penetration must be a number between 0 and 100'
        };
    }
    return { isValid: true };
}

/**
 * Validates a true count value for card counting.
 * @param {unknown} trueCount - The true count to validate
 * @returns {ValidationResult} Validation result object
 * @example
 * // Returns { isValid: true }
 * validateTrueCount(3.5);
 * @example
 * // Returns { isValid: false, message: 'True count must be a number between -20 and 20' }
 * validateTrueCount(25);
 */
function validateTrueCount(trueCount) {
    const count = Number(trueCount);
    if (isNaN(count) || count < -20 || count > 20) {
        return {
            isValid: false,
            message: 'True count must be a number between -20 and 20'
        };
    }
    return { isValid: true };
}

/**
 * Validates a bankroll amount.
 * @param {unknown} bankroll - The bankroll amount to validate
 * @returns {ValidationResult} Validation result object
 * @example
 * // Returns { isValid: true }
 * validateBankroll(1000);
 * @example
 * // Returns { isValid: false, message: 'Bankroll must be a positive number' }
 * validateBankroll(-100);
 */
function validateBankroll(bankroll) {
    const amount = Number(bankroll);
    if (isNaN(amount) || amount <= 0) {
        return {
            isValid: false,
            message: 'Bankroll must be a positive number'
        };
    }
    return { isValid: true };
}

/**
 * Validates minimum and maximum bet amounts.
 * @param {unknown} minBet - The minimum bet amount to validate
 * @param {unknown} maxBet - The maximum bet amount to validate
 * @returns {ValidationResult} Validation result object
 * @example
 * // Returns { isValid: true }
 * validateBetLimits(10, 500);
 * @example
 * // Returns { isValid: false, message: 'Minimum bet must be a positive number' }
 * validateBetLimits(0, 500);
 * @example
 * // Returns { isValid: false, message: 'Minimum bet cannot be greater than maximum bet' }
 * validateBetLimits(100, 50);
 */
function validateBetLimits(minBet, maxBet) {
    const min = Number(minBet);
    const max = Number(maxBet);
    
    // Validate minimum bet
    if (isNaN(min) || min <= 0) {
        return {
            isValid: false,
            message: 'Minimum bet must be a positive number'
        };
    }
    
    // Validate maximum bet
    if (isNaN(max) || max <= 0) {
        return {
            isValid: false,
            message: 'Maximum bet must be a positive number'
        };
    }
    
    // Validate that min is not greater than max
    if (min > max) {
        return {
            isValid: false,
            message: 'Minimum bet cannot be greater than maximum bet'
        };
    }
    
    return { isValid: true };
}

/**
 * Game settings object for validation
 * @typedef {Object} GameSettings
 * @property {CardRank[]} [cards] - Optional array of card ranks
 * @property {CardRank} [dealerCard] - Optional dealer's up card
 * @property {number} [decks] - Optional number of decks in the shoe
 * @property {CountingSystem} [countingSystem] - Optional card counting system
 * @property {number} [penetration] - Optional deck penetration percentage (0-100)
 * @property {number} [trueCount] - Optional true count value
 * @property {number} [bankroll] - Optional bankroll amount
 * @property {number} [minBet] - Optional minimum bet amount
 * @property {number} [maxBet] - Optional maximum bet amount
 */

/**
 * Validates a game settings object with multiple validation rules.
 * @param {unknown} settings - The game settings object to validate
 * @returns {ValidationResult} Validation result object
 * @example
 * // Returns { isValid: true }
 * validateGameSettings({
 *   cards: ['A', 'K'],
 *   decks: 6,
 *   countingSystem: 'hiLo',
 *   penetration: 75
 * });
 * @example
 * // Returns { isValid: false, message: 'Invalid card value: X' }
 * validateGameSettings({ cards: ['A', 'X'] });
 */
function validateGameSettings(settings) {
    // Check if settings is an object
    if (typeof settings !== 'object' || settings === null) {
        return {
            isValid: false,
            message: 'Settings must be an object'
        };
    }
    
    // Validate cards if provided
    if ('cards' in settings) {
        const cardsResult = validateCards(settings.cards);
        if (!cardsResult.isValid) return cardsResult;
    }
    
    // Validate dealer card if provided
    if ('dealerCard' in settings && settings.dealerCard !== undefined) {
        if (!isValidCard(settings.dealerCard)) {
            return {
                isValid: false,
                message: `Invalid dealer card: ${settings.dealerCard}. ` +
                        `Valid values are: ${Array.from(VALID_CARDS).join(', ')}`
            };
        }
    }
    
    // Validate number of decks if provided
    if ('decks' in settings && settings.decks !== undefined) {
        const decksResult = validateDecks(settings.decks);
        if (!decksResult.isValid) return decksResult;
    }
    
    // Validate counting system if provided
    if ('countingSystem' in settings && settings.countingSystem !== undefined) {
        const systemResult = validateCountingSystem(settings.countingSystem);
        if (!systemResult.isValid) return systemResult;
    }
    
    // Validate penetration if provided
    if ('penetration' in settings && settings.penetration !== undefined) {
        const penResult = validatePenetration(settings.penetration);
        if (!penResult.isValid) return penResult;
    }
    
    // Validate true count if provided
    if ('trueCount' in settings && settings.trueCount !== undefined) {
        const countResult = validateTrueCount(settings.trueCount);
        if (!countResult.isValid) return countResult;
    }
    
    // Validate bankroll if provided
    if ('bankroll' in settings && settings.bankroll !== undefined) {
        const bankrollResult = validateBankroll(settings.bankroll);
        if (!bankrollResult.isValid) return bankrollResult;
    }
    
    // Validate bet limits if either min or max bet is provided
    if (('minBet' in settings || 'maxBet' in settings) && 
        (settings.minBet !== undefined || settings.maxBet !== undefined)) {
        const minBet = settings.minBet !== undefined ? settings.minBet : 0;
        const maxBet = settings.maxBet !== undefined ? settings.maxBet : Infinity;
        const betResult = validateBetLimits(minBet, maxBet);
        if (!betResult.isValid) return betResult;
    }
    
    return { isValid: true };
}

// Export all validation functions
export {
    isValidCard,
    validateCards,
    validateDecks,
    validateCountingSystem,
    validatePenetration,
    validateTrueCount,
    validateBankroll,
    validateBetLimits,
    validateGameSettings,
    VALID_CARDS,
    VALID_COUNTING_SYSTEMS
};

File: src/api/utils/validation.py
"""
Validation utilities for the Blackjack API.

This module provides a comprehensive set of validation functions for the Blackjack API,
ensuring consistent validation logic across the application. It handles validation of:
- Card values and hands
- Deck counts and penetration
- Counting systems
- Betting limits and true counts

All validation functions raise appropriate custom exceptions with descriptive error messages
that are suitable for API responses. The module follows a consistent pattern for validation
and error reporting.
"""
from typing import Any, Optional, TypeVar, cast, Union, TYPE_CHECKING
from pydantic import BaseModel

# Lazy imports to avoid circular dependencies
if TYPE_CHECKING:
    from pydantic import FieldValidationInfo
    from ..middleware.error_messages import ErrorMessages

# Import validation constants from constants module
from ..constants import VALID_CARDS, VALID_COUNTING_SYSTEMS

# Constants for validation
MIN_DECKS = 1
MAX_DECKS = 10
MIN_TRUE_COUNT = -20
MAX_TRUE_COUNT = 20

from ..middleware.error_messages import ErrorMessages
from ..middleware.exceptions import (
    BlackjackError,
    InvalidCardError,
    InvalidDeckCountError,
    InvalidCountingSystemError,
    ValidationError,
)

# ValidationError class has been moved to src.api.middleware.exceptions


def _ensure_numeric(value: Any, field_name: str) -> Union[int, float]:
    """Ensure the input is a numeric value.

    Args:
        value: The value to check
        field_name: Name of the field for error messages

    Returns:
        Union[int, float]: The validated numeric value

    Raises:
        ValidationError: If the value is not numeric
    """
    if value is None:
        raise ValidationError(
            message=f"{field_name} is required",
            field=field_name,
            value=value,
            expected="a numeric value",
        )
    try:
        return float(value) if "." in str(value) else int(value)
    except (ValueError, TypeError) as e:
        raise ValidationError(
            message=f"{field_name} must be a number, got {value!r}",
            field=field_name,
            value=value,
            expected="a number",
        ) from e


def _validate_range(
    value: float,
    min_val: float,
    max_val: float,
    field_name: str,
    value_type: str = "value",
) -> float:
    """Validate that a value is within the specified range.

    Args:
        value: The value to validate
        min_val: Minimum allowed value (inclusive)
        max_val: Maximum allowed value (inclusive)
        field_name: Name of the field for error messages
        value_type: Type of the value for error messages (e.g., 'count', 'penetration')

    Returns:
        float: The validated value

    Raises:
        ValidationError: If the value is outside the valid range
    """
    if value < min_val or value > max_val:
        error_msg = f"{field_name} must be between {min_val} and {max_val}, got {value}"
        if value_type != "value":
            error_msg = f"{value_type.capitalize()} must be between {min_val} and {max_val}, got {value}"

        raise ValidationError(
            message=error_msg,
            field=field_name,
            value=value,
            expected=f"a value between {min_val} and {max_val}",
        )
    return value


def validate_card(card: Any) -> str:
    """Validate and normalize a single card value.

    This function validates that the input is a string representing a valid card value
    and normalizes it to uppercase. Valid card values are: 2-10, J, Q, K, A.

    Args:
        card: The card value to validate (will be converted to string)

    Returns:
        str: The validated and normalized card value in uppercase

    Raises:
        InvalidCardError: If the card is not a valid card value

    Example:
        >>> validate_card('k')
        'K'
        >>> validate_card('10')
        '10'
    """
    if card is None:
        raise InvalidCardError(
            message="Card value cannot be None",
            details={"valid_cards": sorted(list(VALID_CARDS))},
        )

    try:
        card_str = str(card).strip().upper()
    except Exception as e:
        raise InvalidCardError(
            message=f"Invalid card value: {card!r}",
            details={
                "value": card,
                "valid_cards": sorted(list(VALID_CARDS)),
                "error": str(e),
            },
        ) from e

    # Special case for 10 (allow '10' or 'T')
    if card_str in ("10", "T"):
        return "10"

    # Check if it's a single character card (A, K, Q, J, or 2-9)
    if len(card_str) == 1 and card_str in VALID_CARDS:
        return card_str

    # If we get here, the card is invalid
    raise InvalidCardError(
        message=f"Invalid card value: {card_str!r}",
        details={"value": card_str, "valid_cards": sorted(list(VALID_CARDS))},
    )


def validate_decks(decks: Any) -> int:
    """Validate the number of decks in play.

    Args:
        decks: The number of decks to validate (will be converted to int)

    Returns:
        int: The validated number of decks as an integer

    Raises:
        InvalidDeckCountError: If the deck count is out of range

    Example:
        >>> validate_decks(4)
        4
        >>> validate_decks('6')
        6
    """
    try:
        decks_num = int(_ensure_numeric(decks, "decks"))
        _validate_range(decks_num, MIN_DECKS, MAX_DECKS, "number of decks")
        return decks_num
    except (ValueError, TypeError) as e:
        raise InvalidDeckCountError(decks, MIN_DECKS, MAX_DECKS) from e


def validate_counting_system(system: Any) -> str:
    """Validate the counting system name.

    Args:
        system: The counting system name to validate (will be converted to string)

    Returns:
        str: The validated counting system name in lowercase

    Raises:
        InvalidCountingSystemError: If the counting system is not recognized

    Example:
        >>> validate_counting_system('hilo')
        'hilo'
    """
    if system is None:
        raise InvalidCountingSystemError(
            system="", message="Counting system is required"
        )

    try:
        system_str = str(system).strip().lower()
    except Exception as e:
        raise InvalidCountingSystemError(
            system=str(system)[:100],  # Truncate long strings
            message=f"Invalid counting system format: {system!r}",
            details={"error": str(e)},
        ) from e

    if not system_str:
        raise InvalidCountingSystemError(
            system=system_str, message="Counting system cannot be empty"
        )

    if system_str not in VALID_COUNTING_SYSTEMS:
        raise InvalidCountingSystemError(
            system=system_str,
            valid_systems=sorted(list(VALID_COUNTING_SYSTEMS)),
            message=f"Unsupported counting system: {system_str}",
        )

    return system_str


def validate_penetration(penetration: Any) -> float:
    """Validate deck penetration percentage.

    Args:
        penetration: The penetration value to validate (0-1)

    Returns:
        float: The validated penetration value between 0 (exclusive) and 1 (inclusive)

    Raises:
        InvalidPenetrationError: If penetration is not between 0 and 1 or is invalid

    Example:
        >>> validate_penetration(0.75)
        0.75
    """
    if penetration is None:
        raise InvalidPenetrationError(
            message="Penetration value is required",
            field="penetration",
            expected="a number between 0 and 1 (exclusive)",
        )

    try:
        penetration_float = float(penetration)
    except (ValueError, TypeError) as e:
        raise InvalidPenetrationError(
            message=f"Penetration must be a number, got {penetration!r}",
            field="penetration",
            value=penetration,
            expected="a number between 0 and 1 (exclusive)",
        ) from e

    if not 0 < penetration_float <= 1:
        raise InvalidPenetrationError(
            message=f"Penetration must be between 0 (exclusive) and 1 (inclusive), got {penetration_float}",
            field="penetration",
            value=penetration_float,
            expected="a value between 0 and 1 (exclusive)",
        )

    return penetration_float


def validate_bet_limits(
    max_bet: Optional[Union[float, int, str]] = None,
    min_bet: Optional[Union[float, int, str]] = None,
) -> tuple[Optional[float], Optional[float]]:
    """Validate betting limits and their relationship.

    Args:
        max_bet: The maximum bet value (can be None, float, int, or string)
        min_bet: The minimum bet value (optional, can be None, float, int, or string)

    Returns:
        tuple[Optional[float], Optional[float]]: A tuple of (validated_min_bet, validated_max_bet)

    Raises:
        InvalidBetLimitsError: If bet limits are invalid or inconsistent

    Example:
        >>> validate_bet_limits(100, 10)
        (10.0, 100.0)
        >>> validate_bet_limits(50, None)
        (None, 50.0)
    """

    def parse_bet(bet: Any, field_name: str) -> Optional[float]:
        if bet is None:
            return None

        try:
            bet_float = float(bet)
        except (ValueError, TypeError) as e:
            raise InvalidBetLimitsError(
                message=f"{field_name} must be a non-negative number, got {bet!r}",
                field=field_name.lower().replace(" ", "_"),
                value=bet,
                expected="a non-negative number",
            ) from e

        if bet_float < 0:
            raise InvalidBetLimitsError(
                message=f"{field_name} must be non-negative, got {bet_float}",
                field=field_name.lower().replace(" ", "_"),
                value=bet_float,
                expected="a non-negative number",
            )

        if not math.isfinite(bet_float):
            raise InvalidBetLimitsError(
                message=f"{field_name} must be a finite number, got {bet_float}",
                field=field_name.lower().replace(" ", "_"),
                value=bet_float,
                expected="a finite number",
            )

        return bet_float

    validated_min = parse_bet(min_bet, "Minimum bet") if min_bet is not None else None
    validated_max = parse_bet(max_bet, "Maximum bet") if max_bet is not None else None

    # If both are provided, validate that min <= max
    if (
        validated_min is not None
        and validated_max is not None
        and validated_min > validated_max
    ):
        raise InvalidBetLimitsError(
            message=f"Minimum bet ({validated_min}) cannot be greater than maximum bet ({validated_max})",
            field="bet_limits",
            value={"min_bet": validated_min, "max_bet": validated_max},
            expected="minimum_bet <= maximum_bet",
        )
        if validated_max < validated_min:
            raise ValidationError(ErrorMessages.INVALID_BET_RANGE)

    return validated_min, validated_max
